src/engine/surface_collision.c

```diff
@@ -574,62 +493,50 @@ f32 find_floor(f32 xPos, f32 yPos, f32 zPos, struct Surface **pfloor) {
     *pfloor = NULL;
 
     if (is_outside_level_bounds(x, z)) {
-        #if PUPPYPRINT_DEBUG
-        collisionTime[perfIteration] += osGetTime()-first;
-        #endif
+#if PUPPYPRINT_DEBUG
+        collisionTime[perfIteration] += osGetTime() - first;
+#endif
         return height;
     }
-
     // Each level is split into cells to limit load, find the appropriate cell.
-    cellX = ((x + LEVEL_BOUNDARY_MAX) / CELL_SIZE) & NUM_CELLS_INDEX;
-    cellZ = ((z + LEVEL_BOUNDARY_MAX) / CELL_SIZE) & NUM_CELLS_INDEX;
-
-    // Check for surfaces belonging to objects.
-    surfaceList = gDynamicSurfacePartition[cellZ][cellX][SPATIAL_PARTITION_FLOORS].next;
-    dynamicFloor = find_floor_from_list(surfaceList, x, y, z, &dynamicHeight);
-
+    cellX = GET_CELL_COORD(x);
+    cellZ = GET_CELL_COORD(z);
     // Check for surfaces that are a part of level geometry.
     surfaceList = gStaticSurfacePartition[cellZ][cellX][SPATIAL_PARTITION_FLOORS].next;
     floor = find_floor_from_list(surfaceList, x, y, z, &height);
-
-    // To prevent the Merry-Go-Round room from loading when Mario passes above the hole that leads
-    // there, SURFACE_INTANGIBLE is used. This prevent the wrong room from loading, but can also allow
-    // Mario to pass through.
-    if (!gFindFloorIncludeSurfaceIntangible) {
-        //! (BBH Crash) Most NULL checking is done by checking the height of the floor returned
-        //  instead of checking directly for a NULL floor. If this check returns a NULL floor
-        //  (happens when there is no floor under the SURFACE_INTANGIBLE floor) but returns the height
-        //  of the SURFACE_INTANGIBLE floor instead of the typical -11000 returned for a NULL floor.
-        if (floor != NULL && floor->type == SURFACE_INTANGIBLE) {
-            floor = find_floor_from_list(surfaceList, x, (s32)(height - 200.0f), z, &height);
-        }
-    } else {
-        // To prevent accidentally leaving the floor tangible, stop checking for it.
-        gFindFloorIncludeSurfaceIntangible = FALSE;
-    }
-
+    if (!gFindFloorExcludeDynamic) {
+        // In the next check, only check for floors higher than the previous check
+        dynamicHeight = height;
+        // Check for surfaces belonging to objects.
+        surfaceList = gDynamicSurfacePartition[cellZ][cellX][SPATIAL_PARTITION_FLOORS].next;
+        dynamicFloor = find_floor_from_list(surfaceList, x, y, z, &dynamicHeight);
+        if (dynamicHeight > height) {
+            floor  = dynamicFloor;
+            height = dynamicHeight;
+        }
+    }
+    // To prevent accidentally leaving the floor tangible, stop checking for it.
+    gFindFloorIncludeSurfaceIntangible = FALSE;
+    gFindFloorExcludeDynamic           = FALSE;
```

Comment:
i feel like there are enough floor check settings where we could have a struct instead to simplify things, or maybe pass a set of flags to these functions?

Reply:
non blocking, conversational

Reply:
Yeah I think flags would be better